src/sysext/sysext.c

```diff
@@ -631,14 +858,65 @@ static int merge_hierarchy(
         if (!f)
                 return log_oom();
 
+        /* Modifying the underlying layers while the overlayfs is mounted is technically undefined, but at
+         * least it won't crash or deadlock, as per the kernel docs about overlayfs:
+         * https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#changes-to-underlying-filesystems */
         r = write_string_file(f, FORMAT_DEVNUM(st.st_dev), WRITE_STRING_FILE_CREATE);
         if (r < 0)
                 return log_error_errno(r, "Failed to write '%s': %m", f);
 
+        /* Store information about workdir path. */
+        if (work_dir) {
+                _cleanup_free_ char *base64_work_dir = NULL;
+                ssize_t base64_size;
+
+                free(f);
+                f = path_join(meta_path, image_class_info[image_class].dot_directory_name, "work_dir");
+                if (!f)
+                        return log_oom();
+
+                /* Paths can have newlines for whatever reason, so better encode them to really get a single
+                 * line file. */
+                base64_size = base64mem(work_dir, strlen(work_dir), &base64_work_dir);
+                if (base64_size < 0)
+                        return log_error_errno(base64_size, "Failed to base64 encode work directory path: %m");
+                r = write_string_file(f, base64_work_dir, WRITE_STRING_FILE_CREATE);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write '%s': %m", f);
+        }
+
         /* Make sure the top-level dir has an mtime marking the point we established the merge */
         if (utimensat(AT_FDCWD, meta_path, NULL, AT_SYMLINK_NOFOLLOW) < 0)
                 return log_error_errno(r, "Failed fix mtime of '%s': %m", meta_path);
 
+        /* Bind mount the meta path as read-only on mutable overlays to avoid accidental modifications of the
+         * contents of meta directory, which could lead to systemd thinking that this hierarchy is not our
+         * mount. */
+        if (work_dir) {
+                _cleanup_fclose_ FILE *proc_self_mountinfo = NULL;
+
+                free(f);
+                f = path_join(overlay_path, image_class_info[image_class].dot_directory_name);
+                if (!f)
+                        return log_oom();
+
+                r = mount_nofollow_verbose(LOG_ERR, f, f, NULL, MS_BIND, NULL);
+                if (r < 0)
+                        return r;
+
+                proc_self_mountinfo = fopen("/proc/self/mountinfo", "re");
+                if (!proc_self_mountinfo) {
+                        r = -errno;
+                        return log_debug_errno(r, "Failed to open /proc/self/mountinfo: %m");
```

Comment:
merge these two lines into:

```
return log_debug_errno(errno, …
```

Reply:
`errno` or `-errno`?

Reply:
log_debug_errno() (and related calls) accept both negative and positive errors, and treat them the same. i.e. they compare with abs() internally. hence, don't bother with `-`.